testonly/mapcontents.go

```diff
@@ -0,0 +1,336 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testonly
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian"
+	"github.com/google/trillian/merkle/hashers"
+)
+
+// ErrInvariant indicates that an invariant check failed, with details in msg.
+type ErrInvariant struct {
+	msg string
+}
+
+func (e ErrInvariant) Error() string {
+	return fmt.Sprintf("Invariant check failed: %v", e.msg)
+}
+
+// MapContents is a complete copy of the map's contents at a particular revision.
+type MapContents struct {
```

Comment:
Is this an in-memory implementation of the MapServer? 
Not sure if it would be useful to have it match the MapServer api with protos and all

Reply:
We could send each request to both implementations and check `proto.Equals` on the results.

Reply:
No, `MapContents` is just a wrapper around a Go `map` to hold a mirror of the expected map contents at a revision.   (It's also pre-existing code, just extracted from the maphammer for re-use here.)